### PR TITLE
Fixed List Slicing in Swift

### PR DESCRIPTION
**Pull request by B-rando1**
### Main Fix
In Swift, the only issue was that the `beg` and `end` bounds would automatically be set to `0` and `list.length` respectively if they weren't given.  I updated the code to check if `step` is negative, and if so set `beg` and `end` to `list.length - 1` and `-1`, respectively, to bring it in line with Python's semantics.  If `step` is a variable, it generates an `if`-statement to do this at runtime, and if `step` is a `litInt`, then it does it at generation-time.

### Changes to `ValData`
In order to get the value of a `litInt`, I added an extra field to `valInt` to `ValData`, so its new definition is:
https://github.com/JacquesCarette/Drasil/blob/282c204c56e627466b903f0e37b1076366193068/code/drasil-gool/lib/GOOL/Drasil/AST.hs#L136

This feels somewhat suboptimal, as `valInt` is `Nothing` unless the value is `litInt`.  Because of smart constructors it's not the end of the world, but I wonder if it would be better to have a new datatype, e.g.
```haskell
data ValDataValue = LitInt Integer | LitBool Bool | ... | ValExpr
```

I do think that this functionality is worth the extra complexity, at least for integers.  For example, we could use it in #3746 to add constant-folding of the `+1`'s that need to take place.

### Test Cases
I switched out the single existing test case for `listSlice` with ten new ones, in order to target every edge case I could think of.  This makes it really easy to see if a language has these edge cases covered, as we can simply run the code and see if the output is the same as the Python output.

### Future Changes
The fix will be similar for Java, C#, and C++.  I might get to them soon, or I might wait  for feedback here about changes to `ValData`.

Contributes to #3809

### CODE REVIEW

**Comment by smiths** [comment]
The change sounds good to me @B-rando1, but I'm not an expert.  :smile:  I do have a thought about your test cases.  It looks like you just print the results and leave it to the human to decide if the tests pass.  There is some cognitive load for this because the human will have to think what does correct output look like.  Is there any way you can automate the check of whether the test cases pass?

**Comment by B-rando1** [comment]
Yes, you are correct @smiths.  Currently the test cases only print their results.  I agree that it would be better to have automated testing.

I'm not very familiar with testing methods.  One way I can think of is to have the makefile run the generated code, and pipe the results into a testing script.  I don't know if there are better options, though.  @balacij or @samm82, do you have any ideas for this?

**Comment by smiths** [comment]
If "proper" testing isn't an option, there are some other potential options, as follows:

- have a "golden" version of the generated code from the list slicing tests in stable and make running the test case generation code part of our CI.  This wouldn't test the answers to running the code, but instead it would test the generated code matches our golden version, as we do for the other artifacts.  @balacij will have to comment on whether this is feasible/desirable
- record in comments the expected output, so at least future developers don't have to reinvent what the answers should be

**Comment by B-rando1** [comment]
@samm82 yeah, the tests in `drasil-code/test` all contain GOOL code that is used to generate code in each of the targets.  In theory, I believe we want that running any of the generated files should produce the same result, and for the test cases I added the Python code is the standard.  I wasn't quite sure what your question was, does that help?

As a side note, I just noticed that `stable/gooltest` seems to be less supported than the rest of stable.  This PR passed CI without me updating stable, and there doesn't seem to be a command to stabilize the GOOL tests either.

File: code/drasil-gool/lib/GOOL/Drasil/LanguageRenderer/SwiftRenderer.hs [hunk 1]
```diff
@@ -942,52 +942,83 @@ swiftIndexFunc l v' = do
...
+  Maybe (SValue r) -> Maybe (SValue r) -> SValue r -> MSBlock r
+swiftListSlice vn vo beg' end' step = do
+  stepV <- zoom lensMStoVS step
+  let needIf = (isNothing beg' || isNothing end') && isNothing (valueInt stepV)
+  bName <- if isJust beg' || not needIf then pure ""
+           else genVarName [] "begIdx"
+  eName <- if isJust end' || not needIf then pure ""
+           else genVarName [] "endIdx"
+  let -- If beg or end aren't given and step isn't a litInt, calculate them at
+      --  and store in a variable.
```
Comment (bmaclach) [inline code/drasil-gool/lib/GOOL/Drasil/LanguageRenderer/SwiftRenderer.hs:997]: Something seems to be missing from this comment, "calculate them at ... and store in a variable"

**Comment by bmaclach** [comment]
> As a side note, I just noticed that `stable/gooltest` seems to be less supported than the rest of stable. This PR passed CI without me updating stable, and there doesn't seem to be a command to stabilize the GOOL tests either.
I guess this is why some of the changes to stable seem unrelated to your changes? stable hadn't been updated in a while?

**Review by bmaclach — DISMISSED** [review]
The changes to GOOL look good to me. I agree that your `ValDataValue` idea would be an improvement, but this seems valuable on its own!

File: code/drasil-gool/lib/GOOL/Drasil/LanguageRenderer/SwiftRenderer.hs (hunk 1 above)
Comment (B-rando1) [reply code/drasil-gool/lib/GOOL/Drasil/LanguageRenderer/SwiftRenderer.hs:997]: Good catch!  It's supposed to say "at runtime".  I'll add that back in.

File: code/drasil-gool/lib/GOOL/Drasil/LanguageRenderer/SwiftRenderer.hs [hunk 2]
```diff
@@ -942,52 +942,80 @@ swiftIndexFunc l v' = do
...
+      computedBeg = valueInt stepV >>= (\s -> if s > 0 then pure $ litInt 0
+                                              else pure $ listSize vo #- litInt 1)
+      computedEnd = valueInt stepV >>= (\s -> if s > 0 then pure $ listSize vo
+                                                       else pure $ litInt (-1))
+      -- Choose the values to use for beg and end based on given information
+      begVal = fromMaybe (fromMaybe (valueOf begVar) computedBeg) beg'
+      endVal = fromMaybe (fromMaybe (valueOf endVar) computedEnd) end'
+      i = var "i" int
+  block $ findBounds ++ [vn &= swiftMapFunc
+    (swiftStrideFunc begVal endVal step) (lambda [i] (listAccess vo (valueOf i)))]
```
Comment (balacij) [inline code/drasil-gool/lib/GOOL/Drasil/LanguageRenderer/SwiftRenderer.hs:1016]: After we spoke about this in person, I jumped down the rabbit hole... it can definitely still be cleaned up (better variable names and comments would go a long way in making it more readable), but I think I've de-duplicated all that I can now.

```haskell
swiftListSlice vn vo beg' end' step = do
  stepV <- zoom lensMStoVS step

  let (setBeg, begVal) = makeSetterVal "begIdx" (valueInt stepV) beg' (litInt 0)    (listSize vo #- litInt 1)
      (setEnd, endVal) = makeSetterVal "endIdx" (valueInt stepV) end' (listSize vo) (litInt (-1))
      i = var "i" int

      setToSlice = vn &= swiftMapFunc (swiftStrideFunc begVal endVal step) (lambda [i] (listAccess vo (valueOf i)))
  block [
      setBeg,
      setEnd,
      setToSlice
    ]
  where
    makeSetterVal _     _       (Just v) _  _  = (emptyStmt, v)
    makeSetterVal vName Nothing _        lb rb = 
      let theVar = var vName int
          theSetter = varDecDef theVar $ inlineIf (step ?> litInt 0) lb rb
      in (theSetter, valueOf theVar)
    makeSetterVal vName (Just s) Nothing  lb rb = 
      let theVar = var vName int
          computed = pure $ if s > 0 then lb else rb
      in (emptyStmt, fromMaybe (valueOf theVar) computed)
```

Here were my steps:
1. Bunch up the begX and endX code by beg and end.
2. Abstract those bunches of code into a local function.
3. Realized a small redundancy in how `needIf` is used, so I inlined it and simplified all expressions.

What do you think?

Comment (balacij) [reply code/drasil-gool/lib/GOOL/Drasil/LanguageRenderer/SwiftRenderer.hs:1016]: Forgot to inline the last `makeSetterVal` case-set...

```haskell
swiftListSlice vn vo beg' end' step = do
  stepV <- zoom lensMStoVS step

  let (setBeg, begVal) = makeSetterVal "begIdx" (valueInt stepV) beg' (litInt 0)    (listSize vo #- litInt 1)
      (setEnd, endVal) = makeSetterVal "endIdx" (valueInt stepV) end' (listSize vo) (litInt (-1))
      i = var "i" int

      setToSlice = vn &= swiftMapFunc (swiftStrideFunc begVal endVal step) (lambda [i] (listAccess vo (valueOf i)))
  block [
      setBeg,
      setEnd,
      setToSlice
    ]
  where
    makeSetterVal _     _       (Just v) _  _  = (emptyStmt, v)
    makeSetterVal vName Nothing _        lb rb = 
      let theVar = var vName int
          theSetter = varDecDef theVar $ inlineIf (step ?> litInt 0) lb rb
      in (theSetter, valueOf theVar)
    makeSetterVal _    (Just s) Nothing  lb rb = (emptyStmt, if s > 0 then lb else rb)
```

Comment (balacij) [reply code/drasil-gool/lib/GOOL/Drasil/LanguageRenderer/SwiftRenderer.hs:1016]: Ok, final time:

```haskell
swiftListSlice vn vo beg end step = do
  stepV <- zoom lensMStoVS step

  let mbStepV = valueInt stepV
      (setBeg, begVal) = makeSetterVal "begIdx" mbStepV beg (litInt 0)    (listSize vo #- litInt 1)
      (setEnd, endVal) = makeSetterVal "endIdx" mbStepV end (listSize vo) (litInt (-1))
      
      i = var "i" int
      setToSlice = vn &= swiftMapFunc (swiftStrideFunc begVal endVal step) (lambda [i] (listAccess vo (valueOf i)))
  block [
      setBeg,
      setEnd,
      setToSlice
    ]
  where
    makeSetterVal _     _       (Just v) _  _  = (emptyStmt, v)
    makeSetterVal _    (Just s) _        lb rb = (emptyStmt, if s > 0 then lb else rb)
    makeSetterVal vName _       _        lb rb = 
      let theVar = var vName int
          theSetter = varDecDef theVar $ inlineIf (step ?> litInt 0) lb rb
      in (theSetter, valueOf theVar)
```

File: code/stable/gooltest/python/PatternTest/Makefile [hunk 3]
```diff
@@ -1,6 +1,10 @@
+# Generated by Drasil v0.1-alpha
+
+# Project Name: PatternTest
+
 build:
 
 run: build
-	python PatternTest.py $(RUNARGS)
+	python3 PatternTest.py $(RUNARGS)
```
Comment (balacij) [inline code/stable/gooltest/python/PatternTest/Makefile:8]: I guess these GOOL diff tests weren't plugged into the CI?

**Review by balacij — DISMISSED** [review]
Everything else looks good to me! Great work!

File: code/drasil-gool/lib/GOOL/Drasil/LanguageRenderer/SwiftRenderer.hs (hunk 2 above)
Comment (B-rando1) [reply code/drasil-gool/lib/GOOL/Drasil/LanguageRenderer/SwiftRenderer.hs:1016]: That looks great to me!  It'll also be handy soon when we make the fix for the `listSlice` in [Macros.hs](https://github.com/JacquesCarette/Drasil/blob/ab8043b24dbd6e5e4664013cf4a52d77d227956c/code/drasil-gool/lib/GOOL/Drasil/LanguageRenderer/Macros.hs#L61-L78), that Java, C#, and C++ use.  We can move `makeSetterVal` to [CommonPseudoOO.hs](https://github.com/JacquesCarette/Drasil/blob/main/code/drasil-gool/lib/GOOL/Drasil/LanguageRenderer/CommonPseudoOO.hs), and then both functions can use it!

Comment (B-rando1) [reply code/drasil-gool/lib/GOOL/Drasil/LanguageRenderer/SwiftRenderer.hs:1016]: I tested it, and it passed all the test cases!  Would you like to push the commit, or should I?

**Review by JacquesCarette — APPROVED** [review]

**Comment by JacquesCarette** [comment]
Looks good. Yes, your suggested refactor of `ValData` would be a good follow-up to this.

File: code/drasil-gool/lib/GOOL/Drasil/LanguageRenderer/SwiftRenderer.hs (hunk 2 above)
Comment (balacij) [reply code/drasil-gool/lib/GOOL/Drasil/LanguageRenderer/SwiftRenderer.hs]: (spoke in person)

**Review by balacij — APPROVED** [review]
Great work!
